### ISSUE TITLE
PyTorch cannot be linked (libtorch_cuda.so) at the last step of compilation with CUDA 11.1 with PyTorch/builder

### ISSUE THREAD

**Issue by UlrichFreitag** [pytorch/pytorch#48588]
## 🐛 Bug

I can't build PyTorch for CUDA 11.1 and cuDNN 8.0.5 using https://github.com/pytorch/builder

## To Reproduce

To be able to compile PyTorch, I have created a small script that uses the functionnality of PyTorch Builder:

`build_pytorch.sh`
```bash
#!/usr/bin/env bash
set -eux

CUDA_VERSION=11.1
DESIRED_PYTHON=3.8m
PYTORCH_VERSION=1.8.0
USE_NIGHTLY="TRUE"
MANYWHEELS_BUILD_DIR=_build

CUDA_VERSION_NO_DOT=$(echo $CUDA_VERSION | tr -d '.')

cd "$(dirname "$0")"  # move inside the script directory
mkdir -p "${MANYWHEELS_BUILD_DIR}"
nvidia-docker run --rm -it \
    --env DESIRED_PYTHON="${DESIRED_PYTHON}" \
    --env PYTORCH_VERSION="${PYTORCH_VERSION}" \
    --env USE_NIGHTLY="${USE_NIGHTLY}" \
    --volume "$(pwd)/${MANYWHEELS_BUILD_DIR}:/remote" \
    --volume "$(pwd)/entrypoint_build.sh:/entrypoint_build.sh" \
    --entrypoint /entrypoint_build.sh \
    "pytorch/manylinux-cuda${CUDA_VERSION_NO_DOT}
```

`entrypoint.sh`
```bash
#!/usr/bin/env bash
set -eux

# setting up git
git config --global user.email "build@pytorch.org"
git config --global user.name "PyTorch Builder"

# cloning pytorch builder
git clone https://github.com/pytorch/builder /builder

# cloning pytorch in the proper place in the docker
if [[ "${USE_NIGHTLY}" == TRUE ]]; then
    git clone --recursive https://github.com/pytorch/pytorch.git /pytorch
else
    git clone -b "v${PYTORCH_VERSION}" --recursive https://github.com/pytorch/pytorch.git /pytorch
fi

# setting env variables for build
export PYTORCH_BUILD_VERSION="${PYTORCH_VERSION}"      # setting env variable for build
if [[ "${USE_NIGHTLY}" == TRUE ]]; then
    export PYTORCH_BUILD_VERSION="${PYTORCH_BUILD_VERSION}+nightly"
fi
export PYTORCH_BUILD_NUMBER=1
export MAX_JOBS=12

cd /builder
manywheel/build.sh
```

This script is quite useful since it allows me to test patch (like Large Model Support) with little change, and build specific version. PyTorch Builder is quite amazing. At the end I just have a wheel that I can use everywhere. So it is quite useful.

However, with CUDA 11.1 (cuDNN 8.0.5 in PyTorch builder https://github.com/pytorch/builder/blob/3e8c5811aa3521674898fb50afee6c8d0ddc8298/common/install_cuda.sh#L140), Python 3.8 it just doesn't link at the end. Here the error that I get:

```
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/codegen/cuda/type.cpp.o
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/tensorexpr/cuda_codegen.cpp.o
[ 92%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/cuda/nccl.cpp.o
[ 92%] Linking CXX shared library ../lib/libtorch_cuda.so
CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/AffineGridGenerator.cpp.o: In function `_GLOBAL__sub_I_AffineGridGenerator.cpp':                                                                            
AffineGridGenerator.cpp:(.text.startup+0x3): relocation truncated to fit: R_X86_64_PC32 against `.bss'
AffineGridGenerator.cpp:(.text.startup+0x21): relocation truncated to fit: R_X86_64_PC32 against `.bss'
CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/RNN.cpp.o: In function `at::native::(anonymous namespace)::get_dropout_state(double, bool, c10::TensorOptions)':
RNN.cpp:(.text+0x13bc): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x13c7): relocation truncated to fit: R_X86_64_PC32 against `.bss'                                                                                                                                   
RNN.cpp:(.text+0x13e7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x140e): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x1415): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x149c): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14a7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14b5): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14f1): additional relocation overflows omitted from the output
collect2: error: ld returned 1 exit status         
gmake[2]: *** [lib/libtorch_cuda.so] Error 1
gmake[1]: *** [caffe2/CMakeFiles/torch_cuda.dir/all] Error 2             
gmake: *** [all] Error 2
Traceback (most recent call last):                                                                        
  File "setup.py", line 765, in <module>
    build_deps()
  File "setup.py", line 315, in build_deps
    build_caffe2(version=version,        
  File "/pytorch/tools/build_pytorch_libs.py", line 58, in build_caffe2                      
    cmake.build(my_env)
  File "/pytorch/tools/setup_helpers/cmake.py", line 346, in build                                     
    self.run(build_args, my_env)                                                                          
  File "/pytorch/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)                                                      
  File "/opt/python/cp38-cp38/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake3', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '12']' returned non-zero exit status 2.
```

The same error happens with PyTorch 1.7.0 (and not just the last version of master). What I don't get is that the CI Build for PyTorch master, CUDA 11.1, cuDNN 8.0.5 pass. Soooo :thinking: 

## Environment

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux): Linux

 - Build command you used (if compiling from source): see above: `./build_pytorch.sh`
 - Python version: 3.8
 - CUDA/cuDNN version: CUDA 11.1 / cuDNN 8.0.5
 - GPU models and configuration: RTX 3090
 - Any other relevant information:

cc @malfet @seemethere @walterddr @ngimel

**Comment by byronyi** [comment]
Try `export TORCH_CUDA_ARCH_LIST="8.6"`.

**Comment by ejguan** [comment]
CUDA11.1 may not be supported due to internal error (https://github.com/pytorch/pytorch/issues/47820#issuecomment-725906418)
#47820 and #47109

**Comment by janeyx99** [comment]
Ah, try cutting down on some of the supported architectures to alleviate this as @byronyi suggested.

This is a problem because the binary for CUDA 11.1 gets large and the relocation markers are only 32 bits. Relocation markers, to my limited understanding, marks offsets to other linked libraries. Since the binary got bigger, the gaps between stuff got bigger as well. Thus, when the marker needs to specify an offset (pinpointing the location of a linked library) that is too large, it overflows. 

I've been running into this issue myself while trying to enable CUDA 11.1 for nightlies and am working on a fix to reorder the nv_fatbin to be after the bss section to mitigate this gap issue.

**Comment by UlrichFreitag** [comment]
> Try `export TORCH_CUDA_ARCH_LIST="8.6"`.

@byronyi @janeyx99 Even in doing so just before calling `manywheel/build.sh` in my script, I still have the same error:

```
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/codegen/cuda/transform_iter.cpp.o
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/codegen/cuda/transform_replay.cpp.o
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/codegen/cuda/transform_rfactor.cpp.o
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/codegen/cuda/type.cpp.o
[ 91%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/jit/tensorexpr/cuda_codegen.cpp.o
[ 92%] Building CXX object caffe2/CMakeFiles/torch_cuda.dir/__/torch/csrc/cuda/nccl.cpp.o
[ 92%] Linking CXX shared library ../lib/libtorch_cuda.so
CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/AffineGridGenerator.cpp.o: In function `_GLOBAL__sub_I_AffineGridGenerator.cpp':
AffineGridGenerator.cpp:(.text.startup+0x3): relocation truncated to fit: R_X86_64_PC32 against `.bss'
AffineGridGenerator.cpp:(.text.startup+0x21): relocation truncated to fit: R_X86_64_PC32 against `.bss'
CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/RNN.cpp.o: In function `at::native::(anonymous namespace)::get_dropout_state(double, bool, c10::TensorOptions)':
RNN.cpp:(.text+0x13bc): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x13c7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x13e7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x140e): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x1415): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x149c): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14a7): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14b5): relocation truncated to fit: R_X86_64_PC32 against `.bss'
RNN.cpp:(.text+0x14f1): additional relocation overflows omitted from the output
collect2: error: ld returned 1 exit status
gmake[2]: *** [lib/libtorch_cuda.so] Error 1
gmake[1]: *** [caffe2/CMakeFiles/torch_cuda.dir/all] Error 2
gmake: *** [all] Error 2
Traceback (most recent call last):
  File "setup.py", line 765, in <module>
    build_deps()
  File "setup.py", line 315, in build_deps
    build_caffe2(version=version,
  File "/pytorch/tools/build_pytorch_libs.py", line 58, in build_caffe2
    cmake.build(my_env)
  File "/pytorch/tools/setup_helpers/cmake.py", line 346, in build
    self.run(build_args, my_env)
  File "/pytorch/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/opt/python/cp38-cp38/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake3', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '12']' returned non-zero exit status 2.
```

**Comment by byronyi** [comment]
@UlrichFreitag Make sure the env doesn't get override like here: https://github.com/pytorch/builder/blob/master/manywheel/build.sh#L52

**Comment by UlrichFreitag** [comment]
@byronyi I used the https://github.com/pytorch/builder/pull/604 feature using the env variable OVERRIDE_TORCH_CUDA_ARCH_LIST and it worked nicely.

@janeyx99 I was wondering however, is this the current blocker to have the official support of PyTorch of CUDA 11.1 in nightlies? If so, do you have a date of when this will be fixed?

**Comment by janeyx99** [comment]
There's no official timeline for getting in CUDA 11.1 support, but we are working closely with the NVIDIA team to get it as soon as possible. We'll notify you when we get more of a real estimate.

**Comment by nelson-liu** [comment]
Is this solved by #49050 , @janeyx99 (sorry for pinging you so much over the last few days)? I tried enabling `BUILD_SPLIT_CUDA` and building 1.8.0 with `pytorch/builder`, but I got:

```
Compiling  functions.cu                        > /pytorch/build/nccl/obj/collectives/device/functions.o
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
nvcc warning : The 'compute_35', 'compute_37', 'compute_50', 'sm_35', 'sm_37' and 'sm_50' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
In file included from /usr/local/cuda/bin/crt/link.stub:129:0:
/tmp/tmpxft_0000152e_00000000-4_devlink.fatbin.c:32419162:0: internal compiler error: Segmentation fault
 ".text\n");

Please submit a full bug report,
with preprocessed source if appropriate.
See <http://bugzilla.redhat.com/bugzilla> for instructions.
Preprocessed source stored into /tmp/ccqlFoDo.out file, please attach this to your bugreport.
make[5]: *** [/pytorch/build/nccl/obj/collectives/device/devlink.o] Error 1
make[4]: *** [/pytorch/build/nccl/obj/collectives/device/colldevice.a] Error 2
make[3]: *** [src.build] Error 2
gmake[2]: *** [nccl_external-prefix/src/nccl_external-stamp/nccl_external-build] Error 2
gmake[1]: *** [CMakeFiles/nccl_external.dir/all] Error 2
gmake: *** [all] Error 2
Building wheel torch-1.8.0+cu111
-- Building version 1.8.0+cu111
cmake3 -DBUILD_PYTHON=True -DBUILD_PYTHONLESS= -DBUILD_SPLIT_CUDA=ON -DBUILD_TEST=True -DCAFFE2_STATIC_LINK_CUDA=1 -DCMAKE_ARGS= -DCMAKE_BUILD_TYPE=Release -DCMAKE_INCLUDE_PATH=/opt/intel/include: -DCMAKE_INSTALL_PREFIX=/pytorch/torch -DCMAKE_LIBRARY_PATH=/opt/intel/lib:/lib: -DCMAKE_PREFIX_PATH=/opt/python/cp36-cp36m/lib/python3.6/site-packages -DGLIBCXX_USE_CXX11_ABI=0 -DINSTALL_TEST=0 -DNUMPY_INCLUDE_DIR=/opt/python/cp36-cp36m/lib/python3.6/site-packages/numpy/core/include -DPYTHON_EXECUTABLE=/opt/python/cp36-cp36m/bin/python -DPYTHON_INCLUDE_DIR=/opt/python/cp36-cp36m/include/python3.6m -DPYTHON_LIBRARY=/opt/_internal/cpython-3.6.6/lib/libpython3.6m.a -DTORCH_BUILD_VERSION=1.8.0+cu111 -DUSE_FBGEMM=0 -DUSE_NUMPY=True -DUSE_STATIC_CUDNN=1 -DUSE_STATIC_NCCL=1 /pytorch
cmake3 --build . --target install --config Release -- -j 12
Traceback (most recent call last):
  File "setup.py", line 818, in <module>
    build_deps()
  File "setup.py", line 320, in build_deps
    cmake=cmake)
  File "/pytorch/tools/build_pytorch_libs.py", line 58, in build_caffe2
    cmake.build(my_env)
  File "/pytorch/tools/setup_helpers/cmake.py", line 345, in build
    self.run(build_args, my_env)
  File "/pytorch/tools/setup_helpers/cmake.py", line 140, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/opt/python/cp36-cp36m/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake3', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '12']' returned non-zero exit status 2.
+ echo 'Stopping container...'
Stopping container...
+ docker rm -f e154d552cdeb69a2d9edcfd59617fe610ed2d10c37383c022c5f3fa3ff92a851
```

**Comment by janeyx99** [comment]
@nelson-liu no worries! Huh, I never saw this error when working with BUILD_SPLIT_CUDA--could you start a new issue with more context? I'm not quite sure what would be causing that error.

**Comment by nelson-liu** [comment]
@janeyx99 I modified my build script to better match the environment variables set in the nightly builds on CI (I was still using an outdated script from the `builder` repo). After making this change, I was able to successfully build 11.1 with `pytorch/builder`, so I'd imagine it was something about a missing / incorrectly set environment variable.

As a result, I think this issue can be closed now. Sorry for the noise, and thanks again for the help!
